### PR TITLE
React Native SDK and Sample Application Improvements

### LINKED ISSUE #10: Make the certificate name configurable
**Describe the issue:**
In the React Native application need to add the certificate to handle SSL issues, Now our SDK only supports the certificate name as a "wso2carbon". Better to improve the SDK to support the certificate name as a name defined by the user.

### LINKED ISSUE #11: Remove the copied files from auth-js and reuse the method
**Describe the issue:**
Now some files do not export from the auth-js SDK, So has been copied those files and used into the React Native SDK. After the files will export, can remove those files and reuse the methods from auth-js.

### LINKED ISSUE #16: Got "[Error: Secure random number generation is not supported by this browser. Use Chrome, Firefox or Internet Explorer 11]"
## Description

I got the error `[Error: Secure random number generation is not supported by this browser. Use Chrome, Firefox or Internet Explorer 11]` when calling the method `getAuthorizationURL` [here](https://github.com/asgardeo/asgardeo-react-native-oidc-sdk/blob/304cb94efd7dc747592729414b58109c9e81f1dd/sample/Screen/LoginScreen.js#L91).

I suspect that the error comes from the method `getCodeVerifier` [here](https://github.com/asgardeo/asgardeo-auth-js-sdk/blob/master/lib/src/utils/crypto-utils.ts) because it uses the package `randombytes`. There is also an issue with the same error https://github.com/crypto-browserify/randombytes/issues/6

### PR DESCRIPTION
**Pull request by ThaminduDilshan**
## Purpose
The sdk and sample application is written to be compatible with a older version of the `auth-js` sdk and had some known bugs. Some of the core sdk functionalities were duplicated in the react native sdk due to unavailability of those at the time of implementation. Moreover some of the functionalities should reside in the sdk rather than the sample application. This PR is intend to fix these issues and update with the latest `auth-js` version.

## Goals
- Fix known issues of the react native sdk and sample application including crypto utils issue.
- Remove duplicated implementations and utilize core sdk functionalities.
- Update react native sdk to be compatible with latest auth-js version (1.0.0).
- Introduce context API to the sdk and maintain authentication state in the sdk.
- Move url redirection to the sdk.
- Introduce context API to maintain application state.

## Release note
- Fix #11 
- Fix #10 
  - Allows to provide CA certificate using the native code.
- Fix #16
  - Updating with auth-js v1.0.0 (Provide crypto utils class implementation to the auth-js)
  - Use `jsrsasign` for jwt operations instead of `jose` library.
- Introduce context API to the sdk
- Handle url redirection from the sdk and expose `signIn` function
- Updated to be compatible with auth-js v1.0.0
- Introduce context API to maintain application state.

| Android version | Device | Device Type | Status |
| ----------- | ----------- | ----------- | ----------- |
| Android 8.0 | Pixel C | Emulator | ✔ |
| Android 8.1 | Pixel 4 XL | Emulator | ✔ |
| Android 9.0 | Galaxy nexus | Emulator | ✔ |
| Android 9.0 | Nexus 10 | Emulator | ✔ |
| Android 10.0 | Pixel 4 | Emulator | ✔ |
| Android 11.0 | Pixel XL | Emulator | ✔ |
| Android 11.0 | Redmi Note 10 | Device | ✔ (release build) |
| Android 11.0 | MI A3 | Device | ✔ (release build) |

### CODE REVIEW

File: sample/App.tsx [hunk 1]
```diff
@@ -20,44 +20,34 @@ import 'react-native-gesture-handler';
...
-          options={{headerShown: false}}
-        />
-        <Stack.Screen
-          name="Auth"
-          component={Auth}
-          options={{headerShown: false}}
-        />
-      </Stack.Navigator>
-    </NavigationContainer>
+    <AuthProvider>
```
Comment (brionmario) [inline sample/App.tsx:33]: Add a rule in ESLint to indent with 4 spaces.

File: sample/context/LoginContext.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    iat: '',
+    iss: '',
+    nbf: '',
+    sub: '',
+    loading: false
+};
+
+const LoginContext = React.createContext();
+
+const LoginContextProvider = (props) => {
```
Comment (brionmario) [inline sample/context/LoginContext.js:45]: Any reason why this file is not written with typescript?

File: sdk/src/wrapper.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
+/**
```
Comment (thivi) [inline sdk/src/wrapper.ts:1]: Can we use a better name for this file?

File: sdk/src/authenticate.tsx [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+                            "AUTHENTICATE-HAR-IV01",
+                            "authenticate",
+                            "handleAuthRedirect",
+                            "Error in signout",
+                            error
+                        )
+                    }
+                }
+            } else {
+                console.log("Invalid redirection url");
```
Comment (thivi) [inline sdk/src/authenticate.tsx:428]: I think it is better to remove all console logs

**Review by thivi — APPROVED** [review]

File: sample/context/LoginContext.js (hunk 2 above)
Comment (ThaminduDilshan) [reply sample/context/LoginContext.js]: Fixed

File: sdk/src/wrapper.ts (hunk 3 above)
Comment (ThaminduDilshan) [reply sdk/src/wrapper.ts:1]: Let's rename it to `store` and move client instantiation logic to `authenticate.tsx`. WDYT?

File: sdk/src/authenticate.tsx (hunk 4 above)
Comment (ThaminduDilshan) [reply sdk/src/authenticate.tsx]: Removed and added `TODO`s to track logging as a future development.

**Review by thivi — DISMISSED** [review]

File: sdk/src/wrapper.ts (hunk 3 above)
Comment (thivi) [reply sdk/src/wrapper.ts]: Yes, let's do this.

**Review by thivi — APPROVED** [review]

**Pull request merged**